### PR TITLE
WIP: add possibility to add animated gif to report

### PR DESCRIPTION
**Pull request by kingjr**
I would like to have the possibility to add animated gif onto the report. I thought it'd be easy by just changing  `image_format`, but that doesn't work. 

The motivation for this EHN is to be able to have animated figures in the report. Another strategy would be to create a play() option to the slider, but I like animated gif... They're easy to save and put everywhere (powerpoint etc).

Any thoughts

### CODE REVIEW

**Comment by mainakjas** [comment]
Can you describe a use case which @teonlamont 's PR does not address?

**Comment by kingjr** [comment]
Overall, it plays automatically, and it's easy export.

In my specific case, I make an [animated figure](http://matplotlib.org/examples/animation/simple_anim.html), as opposed to generating a series of figures.

**Comment by dengemann** [comment]
Exmaple report would be cool.

**Comment by kingjr** [comment]
I don't manage to make this work, I'm asking for help. I changed the file format in the html template and all this, the image is indeed saved in gif, but not as an animated gif.

**Comment by agramfort** [comment]
please share a debug script.

**Comment by mainakjas** [comment]
@kingjr I think you'll have to convert your `gif` to `base64` and embed it in the html.

If that doesn't work, you'll have to get all the `gif` frames, embed them as `png` images like we do in the sliders and play them in loop using javascript. 

If you are up for it, sounds like a nice rainy sunday afternoon hack :)

File: mne/report.py [hunk 1]
```diff
@@ -571,6 +571,14 @@ def _build_html_slider(slices_range, slides_klass, slider_id):
             <img alt=""
              src="data:image/png;base64,{{img}}">
         {{endif}}
+    {{elif image_format == 'gif'}}
+        {{if scale is not None}}
+            <img alt="" style="width:{{width}}%;"
+             src="data:image/gif;base64,{{img}}">
```
Comment (mainakjas) [inline mne/report.py]: does this actually work? Have you tried it using a simple gif ? You should try making this work first with a small python script

**Comment by mainakjas** [comment]
I suspect something is going wrong with your base64 encoding

File: mne/report.py (hunk 1 above)
Comment (kingjr) [inline mne/report.py]: Well, it does put a gif image as wanted. It's just not animated.

Comment (mainakjas) [inline mne/report.py]: which means you just got one frame only ... problem while reading the file then?

**Comment by kingjr** [comment]
> please share a debug script.

You mean something like this?

```
import numpy as np
from matplotlib import animation
import matplotlib.pyplot as plt
from mne.report import Report
report = Report()
fig, ax = plt.subplots(1)
points = ax.scatter(np.random.randn(100), np.random.randn(100), 200)


def animate(nframe):
    for ii in np.linspace(0, 1, 10):
        points.set_facecolors(np.random.rand(100, 3))

anim = animation.FuncAnimation(fig, animate, frames=100)

anim.save('example.gif', writer='imagemagick')
report.add_images_to_section('example.gif', 'section', 'example',
                             image_format='gif')
report.save()
```


> @kingjr I think you'll have to convert your gif to base64 and embed it in the html.

Is this different from what I'm doing now

> If that doesn't work, you'll have to get all the gif frames, embed them as png images like we do in the sliders and play them in loop using javascript.

That seems like an overkill to me. There must be a simpler way.

File: mne/report.py (hunk 1 above)
Comment (mainakjas) [inline mne/report.py]: you should just write 15 lines of python script first which reads a gif, converts to base64 and outputs an html. That will help you debug.

**Comment by mainakjas** [comment]
look at how the images are read in the `report.add_images_to_section` function. I'd recommend first a script without any `mne` at all in it.

**Comment by mainakjas** [comment]
... at some point we should have an example for the `report`. Something that shows all the cool features. And the `report` output html should be linked using some configuration in sphinx (I guess).

**Comment by kingjr** [comment]
Yes, there seems to be a problem with the `Image.save()`. You have to do `imseek(frame)` to get each frame, else it only saves the first one.

I also found this, that may be relevant, https://raw.githubusercontent.com/python-pillow/Pillow/master/Scripts/gifmaker.py but I'm still stuck.. Oh well...

**Comment by agramfort** [comment]
Weather forecast is awful tomorrow I have heard ;)

**Comment by kingjr** [comment]
Now it works, but this should be fixed at PIL level really... But that's beyond my abilities.

File: mne/report.py [hunk 2]
```diff
@@ -887,11 +898,18 @@ def add_images_to_section(self, fnames, captions, scale=None,
...
             # Convert image to binary string.
-            im = Image.open(fname)
             output = BytesIO()
-            im.save(output, format='png')
+            im = Image.open(fname)
+            if image_format == 'png':
+                im.save(output, format='png')
+            elif image_format == 'gif':
+                with open(fname, 'rb') as f:
+                    output.write(f.read())
```
Comment (mainakjas) [inline mne/report.py]: could you do this for all images? Then you don't need the if condition

**Comment by agramfort** [comment]
can you share a demo report?

**Comment by kingjr** [comment]
```
from mne.report import Report
import urllib
urllib.urlretrieve('http://ak-hdl.buzzfed.com/static/enhanced/webdr03/2012' +
                   '/12/19/12/anigif_enhanced-buzz-6068-1355938337-9.gif',
                   'g.gif')

report = Report()
report.add_images_to_section('g.gif', 'section', 'example', image_format='gif')
report.save()
```

https://www.dropbox.com/s/2z2i6emtbb25htb/report.html?dl=0

File: mne/report.py (hunk 2 above)
Comment (kingjr) [inline mne/report.py]: Well I wasn't sure why it was resaved in the first place... Any idea?

**Comment by mainakjas** [comment]
haha cool cc @dengemann @teonlamont ;-)

File: mne/report.py (hunk 2 above)
Comment (dengemann) [inline mne/report.py]: Pretty cool!!

2015-06-28 17:43 GMT+02:00 J-R King <notifications@github.com>:

Comment (mainakjas) [inline mne/report.py]: I think it was to get the binary string but if there's a better way to do it (looks like it), we should probably switch to that.

**Comment by teonlamont** [comment]
how did i do science before this?! :)

File: mne/report.py (hunk 2 above)
Comment (kingjr) [inline mne/report.py]: Actually, I'm thinking that the idea may have been to convert all images into a png format so as to simplify the html part; in which case we could keep like that for now?

**Comment by kingjr** [comment]
More seriously, here's an example to plot an animated gif of a topo to complement @teonlamont's [slider](https://github.com/mne-tools/mne-python/pull/2234).

```
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import animation
import mne
from mne.viz import plot_topomap
from mne.viz.topomap import _prepare_topo_plot

path = mne.datasets.sample.data_path()
fname = path + '/MEG/sample/sample_audvis-ave.fif'
evoked = mne.read_evokeds(fname, condition='Left Auditory', baseline=(None, 0))

sel_times = evoked.times[::10]
fig, ax = plt.subplots(1, figsize=[2, 2])
picks, pos, merge_grads, _, _, = _prepare_topo_plot(evoked, 'mag', None)
txt = ax.text(0., .55, '', ha='center')


def animate(frame):
    time = sel_times[frame]
    time_idx = np.where(evoked.times >= time)[0][0]
    data = evoked.data[picks, time_idx]
    txt.set_text('%i ms' % (1000 * time))
    plot_topomap(data, pos, axis=ax, vmin=-5e-13, vmax=5e-13, sensors=False,
                 contours=False, show=False)

anim = animation.FuncAnimation(fig, animate, frames=len(sel_times),
                               interval=100, repeat=False)
from mne import report
report = report.Report()
report.add_anims_to_section(anim, 'section', 'example')
report.save()
```

This example is higlhly suboptimal for now, as the animation should just change the values of the interpolation rather than replotting everything... I tried, but patience is obviously not my strength... If anyone wants to give it a test.

Also, I'm not very proud of the way I generate the gif. I go through an external file that I pass to `add_images_to_section` and then remove the file.  Note sure about the best alternative to this..

File: example_gif_topo.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        data = evoked.data[picks, time_idx]
+        plot_topomap(data, pos, axis=ax, vmin=-1e-13, vmax=1e-13,
+                     sensors=False, contours=False, show=False)
+
+anim = animation.FuncAnimation(fig, animate, frames=5)
+fig.show()
+
+report = report.Report()
+report.add_anims_to_section(anim, 'section', 'example')
+report.save()
```
Comment (kingjr) [inline example_gif_topo.py]: whoops, this file shouldn't be here, i'll remove it at the next push...

Comment (agramfort) [inline example_gif_topo.py]: please remove any big file from history

**Comment by agramfort** [comment]
can you rebase to see if CIs are happy?

**Pull request closed without merging**